File: src/design_patterns/java/prototype/example/Resume.java
package design_patterns.java.prototype.example;

public class Resume implements Cloneable {
    private String name;
    private String sex;
    private String age;
    private WorkExperience work;

    public Resume(String name) {
        this.name = name;
        this.work = new WorkExperience();
    }

    private Resume(WorkExperience work) {
        this.work = (WorkExperience)work.Clone();
    }

    public void setPersonalInfo(String sex, String age) {
        this.sex = sex;
        this.age = age;
    }

    public void SetWorkExperience(String workDate, String company) {
        work.setWorkDate(workDate);
        work.setWorkCompany(company);
    }

    public void display() {
        System.out.println(this.getName() + " " + this.getSex() + " " + this.getAge());
        System.out.println("Work experience: " + work.getWorkDate() + " " + work.getWorkCompany());
    }

    public Object Clone() {
        Resume obj = new Resume(this.work);
        obj.name = this.name;
        obj.sex = this.sex;
        obj.age = this.age;
        return obj;
    }

    public String getName() {
        return this.name;
    }
    public String getSex() {
        return this.sex;
    }
    public String getAge() {
        return this.age;
    }
}

File: src/design_patterns/cpp/prototype/Resume.cpp
#include<iostream>

#include "Resume.h"

using namespace std;


class WorkExperience: public ICloneable
{
public:
    void setWorkDate(string workDate)
    {
        m_workDate = workDate;
    }

    void setCompany(string company)
    {
        m_company = company;
    }

    string getWorkDate()
    {
        return m_workDate;
    }

    string getCompany()
    {
        return m_company;
    }

    WorkExperience * clone() const override
    {
        return new WorkExperience(*this);
    }

private:
    string m_workDate;
    string m_company;
};


class Resume: public ICloneable
{
public:
    Resume(string name)
    {
        m_name = name;
        m_work = new WorkExperience();
    }

    Resume(WorkExperience * work)
    {
        m_work = work->clone();
    }

    void setPersonalInfo(string sex, string age)
    {
        m_sex = sex;
        m_age = age;
    }

    void setWorkExperience(string workDate, string company)
    {
        m_work->setWorkDate(workDate);
        m_work->setCompany(company);
    }

    void display()
    {
        cout << m_name << " " << m_sex << " " << m_age << endl;
        cout << "工作经历：" << m_work->getWorkDate() << " " << m_work->getCompany() << endl;
    }

    Resume *clone() const override
    {
        Resume *obj = new Resume(m_work);
        obj->m_name = m_name;
        obj->m_sex = m_sex;
        obj->m_age = m_age;
        return obj;
    }

private:
    string m_name;
    string m_sex;
    string m_age;
    WorkExperience * m_work;
};


int main() {

    Resume * a = new Resume("大鸟");
    a->setPersonalInfo("男", "29");
    a->setWorkExperience("1998-2000", "XX公司");

    Resume * b = a->clone();
    b->setWorkExperience("1998-2006", "YY企业");

    Resume * c = a->clone();
    c->setPersonalInfo("男", "24");
    c->setWorkExperience("1998-2003", "ZZ企业");

    a->display();
    b->display();
    c->display();

    return 0;
}